Implement NSAffineTransform

nubbel opened the pull request:
Hi,

The changes in this PR improve the implementation of the transformation methods. The new implementation makes use of the fact that a 2D affine transformation matrix is not an arbitrary 3x3 matrix (see code comments), and is both simpler and more efficient by skipping unnecessary computation steps.
Also included is:
- Implementation of `NSAffineTransformStruct.init()` which initializes all entries to zeros.
- Implementation of `NSAffineTransform.init(transform:)`
- Implementation of `NSCopying` methods for `NSAffineTransform`

jessesquires commented on Foundation/NSAffineTransform.swift:
```diff
@@ -15,7 +15,12 @@ public struct NSAffineTransformStruct {
+    
+    public init() {
+        (self.m11, self.m12, self.m21, self.m22) = (CGFloat(), CGFloat(), CGFloat(), CGFloat())
+        (self.tX, self.tY) = (CGFloat(), CGFloat())
```
this `init` could (should?) just call the full memberwise init below :smile:

nubbel replied:
Ah, you're right. Thanks!

jessesquires commented on Foundation/NSAffineTransform.swift:
```diff
@@ -63,42 +76,30 @@ public class NSAffineTransform : NSObject, NSCopying, NSSecureCoding {
+         [  1 ]     [ 1 ]   [  0    0    1 ] [ 1 ]   [           1        ]
+         */
+        let x = transformStruct.m11*aPoint.x + transformStruct.m12*aPoint.y + transformStruct.tX
+        let y = transformStruct.m21*aPoint.x + transformStruct.m22*aPoint.y + transformStruct.tY
```
i think readability can be improved here with some spacing and `()` :smile: 

```swift
let x = (transformStruct.m11 * aPoint.x) + (transformStruct.m12 * aPoint.y) + transformStruct.tX
```

(same below)

nubbel commented:
Thank you, @jessesquires!
I've incorporated your feedback and updated the PR.

nubbel commented:
I've updated the PR to include the following changes:
- Implementation of translation, scaling and rotation methods with tests
- Implementation for `prependTransform` and `appendTransform` methods

Tests for composed transformations are missing - working on it!
Implementation for `invert` is still in progress, I'm not happy with it yet.

nubbel commented:
I've updated the PR to include the following changes:
- Implementation for `invert`
- Added arithmetic operators for `CGFloat` which were needed for the matrix inversion.

Notes:
- The original implementation throws an exception. My implementation uses `preconditionFailure` in this case.

parkera commented:
Be sure to update the Status doc when we finish a new feature!

https://github.com/apple/swift-corelibs-foundation/blob/master/Docs/Status.md

parkera commented on Foundation/NSAffineTransform.swift:
```diff
@@ -38,67 +51,246 @@ public class NSAffineTransform : NSObject, NSCopying, NSSecureCoding {
+         [   0        0    1 ]
+     */
+    static func rotation(radians angle: CGFloat) -> NSAffineTransformStruct {
+        let α = Double(angle)
```
Love the use of unicode variable name here!

nubbel replied:
Cool, glad you like it. I wasn't sure if that is considered good style, though.

parkera replied:
I think it makes sense to use it in limited places like this, where it actually conveys some meaning.

nubbel commented:
@parkera Sorry, I didn't know about that doc. I updated it in #117.
